web/src/compose_validate.ts

```diff
@@ -795,3 +800,124 @@ export function convert_mentions_to_silent_in_direct_messages(
     const silent_mention_text = people.get_mention_syntax(full_name, user_id, true);
     return silent_mention_text;
 }
+
+export function is_split_messages_enabled(): boolean {
+    return split_messages_enabled;
+}
+
+export function set_split_messages_enabled(enable: boolean): void {
+    split_messages_enabled = enable;
+}
+
+/**
+ * @param content
+ * @returns index of the delimiter occurring outside code blocks since no splitting should
+ * be done inside code blocks
+ */
+function delimiter_index_outside_code_block(content: string): number {
+    let index = content.indexOf(SPLIT_DELIMITER);
+    // skip delimiters which occur inside code blocks
+    while (index !== -1) {
+        if (!compose_ui.position_inside_code_block(content, index + 1)) {
+            return index;
+        }
+        // find next occurrence of the delimiter
+        index = content.indexOf(SPLIT_DELIMITER, index + 1);
+    }
+    return -1;
+}
+
+/**
+ * Does not trim whitespaces that appear just before text since they are
+ * markdown syntax for code blocks.
+ * @param content the string to be trimmed
+ * @returns the trimmed string
+ */
+function trim_except_whitespace_before_text(content: string): string {
+    return content.replace(/^(\n|\s*\n)+/, "").trimEnd();
+}
+
+/**
+ * Splits a message into 2 parts, if possible
+ * @param message_content the content to be split
+ * @returns a pair of strings: the first being the individual split message, the second being the
+ * remaining part of the message, which will be empty when there are no further parts to send
+ */
+export function split_message(raw_message_content: string): [string, string] {
+    if (!is_split_messages_enabled()) {
+        return [raw_message_content, ""];
+    }
+    // trim leading newlines to avoid empty messages when splitting
+    const message_content = trim_except_whitespace_before_text(raw_message_content);
+    const index = delimiter_index_outside_code_block(message_content);
+    if (index === -1) {
+        return [message_content, ""];
+    }
+    const send_content = message_content.slice(0, index);
+    // we need to trim rest_content to remove leading newlines if multiple delimiters are present
+    // in sequence, else it would lead to empty messages when `rest_content` is subsequently passed
+    // to this function for further splitting.
+    const rest_content = trim_except_whitespace_before_text(message_content.slice(index));
+    return [send_content, rest_content];
```

Comment:
I don't understand this comment; doesn't `trim_except_whitespace_before_text` get called at the start of the function too?

Reply:
Yes, it was redundant and I removed it.